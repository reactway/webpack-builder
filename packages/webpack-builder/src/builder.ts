import { Configuration as WebpackConfiguration } from "webpack";

export type Configuration = WebpackConfiguration;
export type UpdateHandler = (webpack: Configuration) => Configuration;
// tslint:disable-next-line no-any
export type Plugin<TConfig = any> = (config: TConfig | undefined, projectDirectory: string) => UpdateHandler;

export class Builder {
    constructor(protected readonly projectDirectory: string, private configuration: Configuration = {}) {}

    public use<TPlugin extends Plugin>(plugin: TPlugin, config?: Parameters<TPlugin>[0]): this {
        this.configuration = plugin(config, this.projectDirectory)(this.configuration);

        return this;
    }

    public update(callback: UpdateHandler): this {
        this.configuration = callback(this.configuration);

        return this;
    }

    public toConfig(enableOptimization?: boolean): Configuration {
        if (this.configuration.entry == null) {
            throw new Error("[Webpack Builder] Entry file is not defined.");
        }

        if (this.configuration.output == null) {
            throw new Error("[Webpack Builder] Output directory is not defined.");
        }

        if (this.configuration.target !== "node" && this.configuration.node == null) {
            this.configuration.node = {
                fs: "empty",
                net: "empty",
                tls: "empty"
            };
        }

        if (enableOptimization === true) {
            this.configuration.optimization = {
                splitChunks: {
                    cacheGroups: {
                        vendors: {
                            test: /[\\/]node_modules[\\/]/,
                            name: "vendor",
                            chunks: "all"
                        }
                    }
                },
                ...this.configuration.optimization
            };
        }

        return this.configuration;
    }
}
