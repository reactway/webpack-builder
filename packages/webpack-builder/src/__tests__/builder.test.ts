__dirname = "TEST_PATH";

jest.mock("upath", () => ({
    ...jest.requireActual("upath"),
    resolve: (...pathSegments: string[]) => pathSegments.join("//"),
    join: (...pathSegments: string[]) => pathSegments.join("//")
}));

import { Configuration } from "webpack";
import upath from "upath";
// Config builder
import { Builder } from "../builder";
// Sample plugin
import { SamplePlugin } from "./sample-plugin/sample-plugin";

let SAMPLE_CONFIGURATION: Configuration = {};
const TEST_PROJECT_LOCATION: string = upath.resolve(__dirname, "./test-project");

beforeEach(() => {
    SAMPLE_CONFIGURATION = {
        entry: "./src/index.ts",
        mode: "development",
        output: {
            path: upath.resolve(TEST_PROJECT_LOCATION, "dist"),
            filename: "[name].bundle.js",
            chunkFilename: "[name].bundle.js",
            publicPath: "./"
        }
    };
});

it("Simple configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION).toConfig();
    expect(configuration).toMatchSnapshot();
});

it("Simple configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION).use(SamplePlugin).toConfig();
    expect(configuration).toMatchSnapshot();
});

it("No configuration given", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION)
        .update(webpack => {
            webpack.entry = SAMPLE_CONFIGURATION.entry;
            webpack.output = SAMPLE_CONFIGURATION.output;
            return webpack;
        })
        .toConfig();
    expect(configuration).toMatchSnapshot();
});

it("Update simple configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION)
        .update(webpack => {
            webpack.mode = "production";
            return webpack;
        })
        .toConfig();

    expect(configuration).toMatchSnapshot();
});

it("No entry file given to configuration", () => {
    const noEntryFileDefined = Object.assign({}, SAMPLE_CONFIGURATION);
    noEntryFileDefined.entry = undefined;
    const configuration = new Builder(TEST_PROJECT_LOCATION, noEntryFileDefined);

    expect(() => configuration.toConfig()).toThrowError("[Webpack Builder] Entry file is not defined.");
});

it("No configuration output given", () => {
    const noOutputDefined = Object.assign({}, SAMPLE_CONFIGURATION);
    noOutputDefined.output = undefined;
    const configuration = new Builder(TEST_PROJECT_LOCATION, noOutputDefined);

    expect(() => configuration.toConfig()).toThrowError("[Webpack Builder] Output directory is not defined.");
});

it("Configuration type node", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION)
        .update(webpack => {
            webpack.target = "node";
            return webpack;
        })
        .toConfig();

    expect(configuration).toMatchSnapshot();
});

it("Simple configuration with optimization", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION).toConfig(true);
    expect(configuration).toMatchSnapshot();
});
