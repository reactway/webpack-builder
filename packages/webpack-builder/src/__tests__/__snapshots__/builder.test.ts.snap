// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration type node 1`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "development",
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
  "target": "node",
}
`;

exports[`No configuration given 1`] = `
Object {
  "entry": "./src/index.ts",
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
}
`;

exports[`Simple configuration 1`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "development",
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
}
`;

exports[`Simple configuration 2`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "loader": "custom-json-loader",
        "test": /\\\\\\.json\\$/,
        "type": "javascript/auto",
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
}
`;

exports[`Simple configuration with optimization 1`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "development",
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendors": Object {
          "chunks": "all",
          "name": "vendor",
          "test": /\\[\\\\\\\\\\\\/\\]node_modules\\[\\\\\\\\\\\\/\\]/,
        },
      },
    },
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
}
`;

exports[`Update simple configuration 1`] = `
Object {
  "entry": "./src/index.ts",
  "mode": "production",
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "TEST_PATH//./test-project//dist",
    "publicPath": "./",
  },
}
`;
